test/loading/index.js

```diff
@@ -0,0 +1,300 @@
+    
+  }
+
+  async init() {
+    this.browser = await puppeteer.launch({
+      headless: true, // change to false for debug
+      defaultViewport: null,
+      ignoreHTTPSErrors: true, 
+      dumpio: false,
+      args: [		
+      '--disable-gpu',
+      '--enable-webgl',
+      '--no-sandbox',
+      '--enable-precise-memory-info',
+      '--enable-begin-frame-control',
+      '--enable-surface-synchronization',
+      '--run-all-compositor-stages-before-draw',
+      '--disable-threaded-animation',
+      '--disable-threaded-scrolling',
+      '--disable-checker-imaging'],
+    });
+    self = this;
+
+    this.page = await this.browser.newPage();
+
+    await this.page.setDefaultNavigationTimeout(60000);
+
+    this.page
+      .on('console', message => {
+        if (message.type().substr(0, 3).toUpperCase() === 'ERR') {
+          this.addStat('ERROR', 'Page-Error: ' + message.text());
+        }
+      })
+      .on('requestfailed', request => {
+        this.addStat('NETWORK', `Request-Error: ${request.failure().errorText} ${request.url()}`);
+      });
+
+    await this.setupRequestInterception();
+
+    await this.page.exposeFunction('PromiseIntercept', this.PromiseIntercept);
+
+      await this.page.evaluateOnNewDocument(() => {
+        Error.stackTraceLimit = 1000;
+        ((Promise) => {
+
+          let originalOnFailure;
+
+          const customFailure = (onFailure)=>{
+              if(typeof onFailure === 'function'){
+                originalOnFailure = onFailure;
+                return customFailureCaller;
+              }else if(typeof originalOnFailure === 'function'){
+                PromiseIntercept(JSON.stringify(onFailure, Object.getOwnPropertyNames(onFailure)));
+                return originalOnFailure(onFailure);
+              }else if(typeof onFailure === 'object'){
+                PromiseIntercept(JSON.stringify(onFailure, Object.getOwnPropertyNames(onFailure)));
+                return onFailure;
+              }
+              return originalOnFailure || onFailure;
+          }
+      
+
+          const customFailureCaller = (e) =>{
+            PromiseIntercept(e);
+            originalOnFailure(e);
+          }
+
+          var originalThen = Promise.prototype.then;
+          Promise.prototype.then = function(onFulfilled, onFailure) {
+             return originalThen.call(this, function(value) {
+              if(onFulfilled){
+                  return onFulfilled(value);   
+              }
+              if(onFailure){
+                originalOnFailure = onFailure;
+              }
+             }, customFailure);
+          };
+      })(this.Promise)
+      })
+  }
+
+  async setupRequestInterception(){
+    await this.page.setRequestInterception(true);
+    this.page.on('request',this.requestListener);
+    this.page.on('requestfailed', this.requestFailedListener);
+    this.page.on('requestfinished', this.requestFinishedListener);
+  }
+
+  async testScene(sceneUrl) {
+    await sleep(5);
```

Comment:
Why the sleep?

Reply:
If there is an actual condition we are waiting for, wait for that. Don't count on it being done in 5 seconds.